File: launch_school/Programming_Foundations/small_problems/lettercase_ratio.rb
=begin
Write a method that takes a string, and then returns a hash that contains 3 entries: one represents the percentage of characters in the string that are lowercase letters, one the percentage of characters that are uppercase letters, and one the percentage of characters that are neither.

**Problem**
-input = string
-output = hash
  keys: lowercase, uppercase, neither
  values: % of lowercase letters, % of uppercase letters, and % of the remainder

clarifications:
-is the input always a string?
-can the input be an empty string?

**Data Structure and Algorithm**
-split the string into characters (stored in array)
-loop through the string using the method count and test each element
=end

def letter_percentages(str)
  total_length = str.size
  lowercase = Rational((str.count("\\a-z") * 100), total_length).to_f.round(2)
  uppercase = Rational((str.count("\\A-Z") * 100), total_length).to_f.round(2)
  neither = 100 - lowercase - uppercase

  { lowercase: lowercase, uppercase: uppercase, neither: neither}
end

p letter_percentages('abCdef 123') == { lowercase: 50, uppercase: 10, neither: 40 }
p letter_percentages('AbCd +Ef') == { lowercase: 37.5, uppercase: 37.5, neither: 25 }
p letter_percentages('123') == { lowercase: 0, uppercase: 0, neither: 100 }

File: launch_school/LS215-Computational-Thinking/small_problems/medium2/lettercase_ratio.js
'use strict';

function letterPercentages1(string) {
  const percentageObj = function (lowercase, uppercase, neither) {
    return { lowercase, uppercase, neither };
  };

  let lowercase = 0;
  let uppercase = 0;
  let neither = 0;
  let total = 0;

  [...string].forEach(char => {
    total += 1;

    if (/[a-z]/.test(char)) {
      lowercase += 1;
    } else if (/[A-Z]/.test(char)) {
      uppercase += 1;
    } else {
      neither += 1;
    }
  });

  const convertPercentage = (number) => {
    return ((number / total) * 100).toFixed(2);
  };

  [lowercase, uppercase, neither] =
    [lowercase, uppercase, neither].map(convertPercentage);

  return percentageObj(lowercase, uppercase, neither);
}

function letterPercentages(string) {
  const percentageObj = function (lowercase, uppercase, neither) {
    return { lowercase, uppercase, neither };
  };

  let lowercase = (string.match(/[a-z]/g) || []).length;
  let uppercase = (string.match(/[A-Z]/g) || []).length;
  let total = string.length;
  let neither = total - lowercase - uppercase;

  const convertPercentage = (number) => {
    return ((number / total) * 100).toFixed(2);
  };

  [lowercase, uppercase, neither] =
    [lowercase, uppercase, neither].map(convertPercentage);

  return percentageObj(lowercase, uppercase, neither);
}

console.log(letterPercentages('abCdef 123'));
// { lowercase: "50.00", uppercase: "10.00", neither: "40.00" }

console.log(letterPercentages('AbCd +Ef'));
// { lowercase: "37.50", uppercase: "37.50", neither: "25.00" }

console.log(letterPercentages('123'));
// { lowercase: "0.00", uppercase: "0.00", neither: "100.00" }
